code/datums/components/fullauto.dm

```diff
@@ -0,0 +1,277 @@
+	mouse_status = AUTOFIRE_MOUSEUP //In regards to the component there's no click anymore to care about.
+	clicker = null
+	if(!QDELETED(shooter))
+		UnregisterSignal(shooter, COMSIG_MOB_LOGOUT)
+	UnregisterSignal(parent, list(COMSIG_PARENT_PREQDELETED, COMSIG_ITEM_DROPPED))
+	shooter = null
+	parent.UnregisterSignal(parent, COMSIG_AUTOFIRE_SHOT)
+	parent.UnregisterSignal(src, COMSIG_AUTOFIRE_ONMOUSEDOWN)
+
+
+/datum/component/automatic_fire/proc/on_mouse_down(client/source, atom/_target, turf/location, control, params)
+	var/list/modifiers = params2list(params) //If they're shift+clicking, for example, let's not have them accidentally shoot.
+	if(modifiers["shift"] && (world.time <= source.mob.next_click || source.mob.ShiftClickOn(_target)))
+		source.click_intercept_time = world.time
+		return
+	if(modifiers["ctrl"])
+		return
+	if(modifiers["middle"])
+		return
+	if(modifiers["alt"] && (world.time <= source.mob.next_click || source.mob.AltClickOn(_target)))
+		source.click_intercept_time = world.time
+		return
+
+	if(source.mob.in_throw_mode)
+		return
+	if(!isturf(source.mob.loc)) //No firing inside lockers and stuff.
+		return
+	if(get_dist(source.mob, _target) < 2) //Adjacent clicking.
+		return
+
+	if(isnull(location)) //Clicking on a screen object.
+		if(_target.plane != CLICKCATCHER_PLANE) //The clickcatcher is a special case. We want the click to trigger then, under it.
+			return //If we click and drag on our worn backpack, for example, we want it to open instead.
+		_target = params2turf(modifiers["screen-loc"], get_turf(source.eye), source)
+		if(!_target)
+			CRASH("Failed to get the turf under clickcatcher")
+
+	if(SEND_SIGNAL(src, COMSIG_AUTOFIRE_ONMOUSEDOWN, source, _target, location, control, params) & COMPONENT_AUTOFIRE_ONMOUSEDOWN_BYPASS)
+		return
+
+	source.click_intercept_time = world.time //From this point onwards Click() will no longer be triggered.
+
+	if(autofire_stat & (AUTOFIRE_STAT_IDLE))
+		CRASH("on_mouse_down() called with [autofire_stat] autofire_stat")
+	if(autofire_stat & AUTOFIRE_STAT_FIRING)
+		stop_autofiring() //This can happen if we click and hold and then alt+tab, printscreen or other such action. MouseUp won't be called then and it will keep autofiring.
+
+	target = _target
+	target_loc = get_turf(target)
+	mouse_parameters = params
+	start_autofiring()
+
+
+//Dakka-dakka
+/datum/component/automatic_fire/proc/start_autofiring()
+	if(autofire_stat == AUTOFIRE_STAT_FIRING)
+		return //Already pew-pewing.
+	autofire_stat = AUTOFIRE_STAT_FIRING
+
+	if(auto_delay_timer) //This shouldn't be happening, so let's stack_trace it and remove it if nothing is caught.
+		stack_trace("start_autofiring called with a non-null auto_delay_timer")
+		if(!deltimer(auto_delay_timer))
+			addtimer(CALLBACK(src, .proc/keep_trying_to_delete_timer, auto_delay_timer), 1)
+		auto_delay_timer = null
+
+	clicker.mouse_override_icon = 'icons/effects/mouse_pointers/weapon_pointer.dmi'
```

Comment:
It's a little weird to have the mouse icon only show up when you're firing

Reply:
wdym i think it works fine